### ISSUE TITLE
Move Product Error from MySQL database

### ISSUE THREAD

**Issue by LaansDole** [miketvo/rmit-isys2099-group9-app#69]
## Add debugging code to move_product PROCEDURE:

```sql
DROP PROCEDURE IF EXISTS sp_move_product;
DELIMITER $$
CREATE PROCEDURE sp_move_product(
    IN move_product INT,
    IN move_quantity INT,
    IN from_warehouse INT,
    IN to_warehouse INT,
    OUT result INT
)
this_proc:
BEGIN
    DECLARE _rollback BOOL DEFAULT 0;
    DECLARE CONTINUE HANDLER FOR SQLEXCEPTION SET _rollback = 1;
    START TRANSACTION;
    SET result = 0;
    SELECT 'result after initialization:', result;

    -- Checks for early termination
    IF move_quantity < 1 THEN SET result = 2; ROLLBACK; LEAVE this_proc; END IF;
    SELECT 'result after checking move_quantity:', result;

    IF from_warehouse = to_warehouse THEN SET result = 2; ROLLBACK; LEAVE this_proc; END IF;
    SELECT 'result after checking from_warehouse and to_warehouse:', result;

    SELECT count(*) INTO @exist_product FROM product WHERE id = move_product FOR SHARE ;
    SELECT '@exist_product:', @exist_product;
    IF @exist_product = 0 THEN SET result = 1; ROLLBACK; LEAVE this_proc; END IF;

    SELECT count(*) INTO @exist_from_warehouse FROM warehouse WHERE id = from_warehouse FOR SHARE;
    SELECT '@exist_from_warehouse:', @exist_from_warehouse;
    IF @exist_from_warehouse = 0 THEN SET result = 1; ROLLBACK; LEAVE this_proc; END IF;

    SELECT count(*) INTO @exist_to_warehouse FROM warehouse WHERE id = to_warehouse FOR SHARE;
    SELECT '@exist_to_warehouse:', @exist_to_warehouse;
    IF @exist_to_warehouse = 0 THEN SET result = 1; ROLLBACK; LEAVE this_proc; END IF;

    SELECT count(*) INTO @from_warehouse_has_product FROM stockpile WHERE product_id = move_product AND warehouse_id = from_warehouse FOR UPDATE;
    SELECT '@from_warehouse_has_product:', @from_warehouse_has_product;
    IF @from_warehouse_has_product = 0 THEN SET result = 1; ROLLBACK; LEAVE this_proc; END IF;

    SELECT quantity INTO @from_warehouse_product_quantity FROM stockpile WHERE product_id = move_product AND warehouse_id = from_warehouse FOR UPDATE;
    SELECT '@from_warehouse_product_quantity:', @from_warehouse_product_quantity;
    IF move_quantity > @from_warehouse_product_quantity THEN SET result = 1; ROLLBACK; LEAVE this_proc; END IF;

    -- Calculate available space in to_warehouse
    SELECT w.volume - coalesce(sum(s.quantity * p.width * p.length * p.height), 0)
    INTO @to_warehouse_available_volume
    FROM stockpile s
        LEFT JOIN warehouse w ON s.warehouse_id = w.id
        LEFT JOIN product p on s.product_id = p.id
    WHERE w.id = to_warehouse FOR SHARE;
    
     -- Calculate the amount of space needed in to_warehouse
     SELECT move_quantity * width * length * height
     INTO @move_volume
     FROM product WHERE id = move_product FOR SHARE;

     SELECT '@to_warehouse_available_volume:', @to_warehouse_available_volume, '@move_volume:', @move_volume;
     IF @to_warehouse_available_volume < @move_volume THEN SET result = 1; ROLLBACK; LEAVE this_proc; END IF;


     -- Update the database for the move
     IF move_quantity = @from_warehouse_product_quantity THEN
         DELETE FROM stockpile WHERE product_id = move_product AND warehouse_id = from_warehouse;
     ELSE
         UPDATE stockpile SET quantity = quantity - move_quantity WHERE product_id = move_product AND warehouse_id = from_warehouse;
     END IF;

     SELECT count(*) INTO @to_warehouse_has_product FROM stockpile WHERE product_id = move_product AND warehouse_id = to_warehouse FOR UPDATE;
     SELECT '@to_warehouse_has_product:', @to_warehouse_has_product;
     IF @to_warehouse_has_product = 0 THEN
         INSERT INTO stockpile (product_id, warehouse_id, quantity) VALUES (move_product, to_warehouse, move_quantity);
     ELSE
         UPDATE stockpile SET quantity = quantity + move_quantity WHERE product_id = move_product AND warehouse_id = to_warehouse;
     END IF;


     -- Commit or Rollback
     IF _rollback THEN
         SET result = -1;
         ROLLBACK;
     ELSE
         COMMIT;
     END IF;

     SELECT 'final result:', result;
END $$
DELIMITER ;
```
## Then
```CALL sp_move_product(1, 1, 1, 3, @result);```

| result after initialization: | result |
|------------------------------|--------|
| result after initialization: |      0 |

| result after checking move_quantity: | result |
|--------------------------------------|--------|
| result after checking move_quantity: |      0 |

| result after checking from_warehouse and to_warehouse: | result |
|--------------------------------------------------------|--------|
| result after checking from_warehouse and to_warehouse: |      0 |

| @exist_product: | @exist_product |
|-----------------|----------------|
| @exist_product: |              1 |

| @exist_from_warehouse: | @exist_from_warehouse |
|------------------------|-----------------------|
| @exist_from_warehouse: |                     1 |

| @exist_to_warehouse: | @exist_to_warehouse |
|----------------------|---------------------|
| @exist_to_warehouse: |                   1 |

| @from_warehouse_has_product: | @from_warehouse_has_product |
|------------------------------|-----------------------------|
| @from_warehouse_has_product: |                           1 |

| @from_warehouse_product_quantity: | @from_warehouse_product_quantity |
|-----------------------------------|----------------------------------|
| @from_warehouse_product_quantity: |                              100 |

| @to_warehouse_available_volume: | @to_warehouse_available_volume | @move_volume: | @move_volume |
|---------------------------------|----------------------------------|-----------------|-----------------|
| @to_warehouse_available_volume: | NULL                                                           | @move_volume: |           72 |

| @to_warehouse_has_product: | @to_warehouse_has_product |
|----------------------------|---------------------------|
| @to_warehouse_has_product: |                         0 |

| final result:             | result |
|-                          |-       |
| final result:             |     -1 |
## Issue
It appears that an SQL exception occurred when calculating the available space in the destination warehouse (to_warehouse). This is indicated by the value of @to_warehouse_available_volume being NULL. This value is calculated by the following statement in your stored procedure:
```sql
SELECT w.volume - coalesce(sum(s.quantity * p.width * p.length * p.height), 0)
INTO @to_warehouse_available_volume
FROM stockpile s
    LEFT JOIN warehouse w ON s.warehouse_id = w.id
    LEFT JOIN product p on s.product_id = p.id
WHERE w.id = to_warehouse FOR SHARE;
```
This statement calculates the available volume in the destination warehouse by subtracting the total volume of all products in the warehouse from the warehouse’s total volume. If an SQL exception occurs when executing this statement, it could cause the value of @to_warehouse_available_volume to be NULL.

There are several reasons why an SQL exception might occur when executing this statement. Some possible causes include:

- The `warehouse` table does not have a row with an id equal to the value of to_warehouse.
- The `product` table does not have rows with ids that match the product_ids in the `stockpile` table.
- There are issues with data types or data conversion when calculating the product volumes or the available volume.

**Comment by miketvo** [comment]
I can't reproduce your result on my local instance of MySQL. I'm using the same mock data you have on `feature-resolve-TODOs-BE`:

```CALL sp_move_product(1, 1, 1, 3, @result);```

| result after initialization: | result |
|------------------------------|--------|
| result after initialization: |      0 |

| result after checking move_quantity: | result |
|--------------------------------------|--------|
| result after checking move_quantity: |      0 |

| result after checking from_warehouse and to_warehouse: | result |
|--------------------------------------------------------|--------|
| result after checking from_warehouse and to_warehouse: |      0 |

| @exist_product: | @exist_product |
|-----------------|----------------|
| @exist_product: |              1 |

| @exist_from_warehouse: | @exist_from_warehouse |
|------------------------|-----------------------|
| @exist_from_warehouse: |                     1 |

| @exist_to_warehouse: | @exist_to_warehouse |
|----------------------|---------------------|
| @exist_to_warehouse: |                   1 |

| @from_warehouse_has_product: | @from_warehouse_has_product |
|------------------------------|-----------------------------|
| @from_warehouse_has_product: |                           1 |

| @from_warehouse_product_quantity: | @from_warehouse_product_quantity |
|-----------------------------------|----------------------------------|
| @from_warehouse_product_quantity: |                              100 |

| @to_warehouse_available_volume: | @to_warehouse_available_volume | @move_volume: | @move_volume |
|---------------------------------|----------------------------------|-----------------|-----------------|
| @to_warehouse_available_volume: | 1872280                               | @move_volume: |           72 |

| @to_warehouse_has_product: | @to_warehouse_has_product |
|----------------------------|---------------------------|
| @to_warehouse_has_product: |                         0 |

| final result:             | result |
|-                          |-       |
| final result:             |     0 |

**Comment by miketvo** [comment]
Potential point of issue: Different MySQL server versions:

- 8.0.28 on Nhat's computer.
- 8.0.29 on Mike's computer.
- 8.0.33 on Loi's and Tony's computers.

Tri will use 8.0.33, so let's upgrade to that and test again.

**Comment by miketvo** [comment]
Confirmed that this bug only appears on 8.0.33, and not 8.0.28 nor 8.0.29. Weird.

**Comment by miketvo** [comment]
Fixed in #73
